1.2.x [ELY-1526] Add back the ability for security providers to be loaded using the service loader mechanism but ensure that we skip any provider that is already an installed provider

fjuma opened the pull request:
https://issues.jboss.org/browse/ELY-1526
https://issues.jboss.org/browse/WFLY-9899

dmlloyd commented on src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java:
```diff
@@ -0,0 +1,86 @@
+ */
+public class ProviderServiceLoaderSupplier extends ServiceLoaderSupplier<Provider> {
+
+    private static final HashSet INSTALLED_PROVIDERS_SET;
```
This raw type should have a type of ``HashSet<Class<?>>`` or ``Set<Class<?>>``.

Also it should not be static.  The set of installed providers may change over time, so the set of types of providers can also change over time.  Therefore the set should be reconstructed each time the ``INSTALLED_PROVIDERS`` supplier is evaluated via its ``get()`` method.

fjuma replied:
Good catch, will fix this.

dmlloyd commented on src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java:
```diff
@@ -0,0 +1,86 @@
+        Iterator<Provider> iterator = loader.iterator();
+        for (;;) try {
+            if (! iterator.hasNext()) {
+                return list.toArray(Arrays2.createArray(service, list.size()));
```
I don't remember if @darranl has a different recommendation here, but AFAIK accepted best practice is to pass in a constant 0-sized array to ``Collection.<T>toArray(T[] array)`` methods.

fjuma replied:
This is actually just a copy/paste from ```ServiceLoaderSupplier#loadServices```. This isn't always 0-sized, this is the result after iterating over the providers.

dmlloyd replied:
``ServiceLoaderSupplier`` is generic; that's why it uses this array creation strategy.  But since this class is not generic, it does not need to use the dynamic creation strategy.  That said, this isn't incorrect, it's just non-optimal.

fjuma replied:
Ah, yes, that was silly :) I'll fix that.

dmlloyd commented on src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java:
```diff
@@ -0,0 +1,86 @@
+        }
+    }
+
+    public int hashCode() {
```
The ``hashCode`` impl should match the ``equals`` impl, that is, it should return the hash code of ``classLoader`` or a value derived directly from it.

fjuma replied:
```super.hashCode()``` does currently return a value derived from the ```classLoader```:

https://github.com/wildfly-security/wildfly-elytron/blob/master/src/main/java/org/wildfly/security/util/ServiceLoaderSupplier.java#L82-L89

Does that look ok?

dmlloyd replied:
Yeah that looks OK.

dmlloyd commented on src/main/java/org/wildfly/security/util/ServiceLoaderSupplier.java:
```diff
@@ -39,7 +39,7 @@
 
     private final Class<E> service;
-    private final ClassLoader classLoader;
+    protected final ClassLoader classLoader;
```
This is an API change; I think we may want to keep this package-private unless we have a need to open it to external subclasses.

dmlloyd commented on src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java:
```diff
@@ -126,9 +128,11 @@
-            new ServiceLoaderSupplier<>(Provider.class, ElytronXmlParser.class.getClassLoader());
+    private static final Supplier<Provider[]> ELYTRON_PROVIDER_SUPPLIER = ProviderUtil.aggregate(
+            () -> new Provider[] { new WildFlyElytronProvider() },
+            WildFlySecurityManager.isChecking() ?
```
I don't think this needs to be privileged since the provider supplier doesn't seem to capture the ACC on construction, unless it does and I just can't see it in this diff...

fjuma replied:
This was switched over to being privileged in ELY-1349: https://github.com/wildfly-security/wildfly-elytron/pull/972

fjuma replied:
The parent class, ```ServiceLoaderSupplier```, captures the ACC on construction.

dmlloyd replied:
OK that makes sense.

fjuma commented:
Updated based on feedback from David.

dmlloyd commented on src/main/java/org/wildfly/security/util/ServiceLoaderSupplier.java:
```diff
@@ -65,7 +65,7 @@ public ServiceLoaderSupplier(final Class<E> service, final ClassLoader classLoad
     }
 
-    private E[] loadServices(final Class<E> service, final ClassLoader classLoader) {
+    protected E[] loadServices(final Class<E> service, final ClassLoader classLoader) {
```
I think we should not make this ``protected``; it would needlessly increase our API surface area (and there's no guarantee we wouldn't want to refactor at this level).  Package-private is probably the better choice here.

fjuma commented:
Updated based on the latest feedback from @dmlloyd and also made the following two changes to address the two issues @mchoma had mentioned on WFLY-9899:

* Updated ```AuthenticationConfiguration#useProvidersFromClassLoader``` to ensure ```ProviderServiceLoaderSupplier``` is used

* Updated handling of ```use-service-loader" in ```ElytronXmlParser``` to ensure ```ProviderServiceLoaderSupplier``` is used

fjuma commented:
Updated ```ProviderServiceLoaderSupplier``` to only treat the ```WildFlyElytronProvider``` as part of the list of installed providers if we know it's been added explicitly:

hkalina commented:
@dmlloyd I should ask you yet, is merging ok, as @fjuma say she has added small second commit yet.

dmlloyd commented:
Yeah still looks OK.

fjuma commented:
Thanks @dmlloyd and @hkalina.

mchoma commented on src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java:
```diff
@@ -39,14 +39,23 @@
  */
 public class ProviderServiceLoaderSupplier extends ServiceLoaderSupplier<Provider> {
 
+    final boolean elytronProviderStaticallyAdded;
```
Why was that necessary to introduce?

hkalina replied:
For cases when will be Elytron provider added AFTER - like https://github.com/wildfly-security/wildfly-elytron/pull/1101/files#diff-c326ff349cb9d8a1a54a2f94e9e3683aL167

fjuma replied:
@mchoma This was added because ```ProviderServiceLoaderSupplier``` is now also being used from ```AuthenticationConfiguration#useProvidersFromClassLoader``` and in the handling of ```<use-service-loader module-name=...>```. In these cases, we shouldn't just filter out ```WildFlyElytronProvider``` if it is found since it wasn't already added explicitly by us like it is for the ```DEFAULT_PROVIDER_SUPPLIER```.

mchoma commented on src/main/java/org/wildfly/security/auth/client/AuthenticationConfiguration.java:
```diff
@@ -168,8 +168,8 @@
     private static final Supplier<Provider[]> DEFAULT_PROVIDER_SUPPLIER = ProviderUtil.aggregate(
             () -> new Provider[] { new WildFlyElytronProvider() },
```
With elytronProviderStaticallyAdded=true this is redundant?

mchoma commented on src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java:
```diff
@@ -130,8 +130,8 @@
     private static final Supplier<Provider[]> ELYTRON_PROVIDER_SUPPLIER = ProviderUtil.aggregate(
             () -> new Provider[] { new WildFlyElytronProvider() },
```
With elytronProviderStaticallyAdded=true this is redundant?
